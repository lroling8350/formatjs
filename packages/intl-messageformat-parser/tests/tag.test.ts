import {pegParse as _pegParse, IParseOptions} from '../src/parser';

// Make it easier to figure out the error location
function pegParse(
  source: string,
  options?: IParseOptions
): ReturnType<typeof _pegParse> {
  try {
    return _pegParse(source, options);
  } catch (err) {
    if (err.location) {
      const {start, end} = err.location;
      err.message = `(${start.line}:${start.column} - ${end.line}:${end.column}) ${err.message}`;
    }
    throw err;
  }
}

test('tag with number arg', () => {
  expect(
    pegParse('I have <foo>{numCats, number}</foo> cats.')
  ).toMatchSnapshot();
});

test('tag with number arg with ignoreTag', () => {
  expect(
    pegParse('I have <foo>{numCats, number}</foo> cats.', {ignoreTag: true})
  ).toMatchSnapshot();
});

test('tag with rich arg', () => {
  expect(
    pegParse(
      'I <b>have</b> <foo>{numCats, number} some string {placeholder}</foo> cats.'
    )
  ).toMatchSnapshot();
});

test('tag with rich arg with ignoreTag', () => {
  expect(
    pegParse(
      'I <b>have</b> <foo>{numCats, number} some string {placeholder}</foo> cats.',
      {ignoreTag: true}
    )
  ).toMatchSnapshot();
});

test('escaped tag with rich arg', () => {
  expect(pegParse("I '<3 cats.")).toMatchSnapshot();
});

test('unescaped left angle bracket', () => {
  expect(pegParse('I <3 cats.')).toMatchSnapshot();
});

test('escaped multiple tags', () => {
  expect(pegParse("I '<'3 cats. '<a>foo</a>' '<b>bar</b>'")).toMatchSnapshot();
});

test('escaped multiple tags with placeholder', () => {
  expect(
    pegParse("I '<'3 cats. '<a>foo</a>' '<b>'{bar}'</b>'")
  ).toMatchSnapshot();
});

test('mismatched tag', function () {
  expect(() => pegParse('this is <a>mismatch</b>')).toThrowError(/Mismatch/);
});

test('mismatched tag with ignoreTag', function () {
  expect(
    pegParse('this is <a>mismatch</b>', {ignoreTag: true})
  ).toMatchSnapshot();
});

test('nested tag', function () {
  expect(
    pegParse('this is <a>nested <b>{placeholder}</b></a>')
  ).toMatchSnapshot();
});

test('nested tag with ignoreTag', function () {
  expect(
    pegParse('this is <a>nested <b>{placeholder}</b></a>', {ignoreTag: true})
  ).toMatchSnapshot();
});

test('tag in plural', function () {
  expect(
    pegParse(
      'You have {count, plural, =1 {<b>1</b> Message} other {<b>#</b> Messages}}'
    )
  ).toMatchSnapshot();
});

test('tag in plural with ignoreTag', function () {
  expect(
    pegParse(
      'You have {count, plural, =1 {<b>1</b> Message} other {<b>#</b> Messages}}',
      {ignoreTag: true}
    )
  ).toMatchSnapshot();
});

test('self-closing tag', function () {
  expect(
    pegParse('this is <br/> <a>nested <b>{placeholder}</b></a>')
  ).toMatchSnapshot();
});

test('self-closing tag with ignoreTag', function () {
  expect(
    pegParse('this is <br/> <a>nested <b>{placeholder}</b></a>', {
      ignoreTag: true,
    })
  ).toMatchSnapshot();
});

test('tag with dash', function () {
  expect(
    pegParse('this is <br/> <dash-tag>nested <b>{placeholder}</b></dash-tag>')
  ).toMatchSnapshot();
});

test('tag with dash with ignoreTag', function () {
  expect(
    pegParse('this is <br/> <dash-tag>nested <b>{placeholder}</b></dash-tag>', {
      ignoreTag: true,
    })
  ).toMatchSnapshot();
});

// https://github.com/formatjs/formatjs/issues/1845
test('issue #1845: less than sign without escape', () => {
  expect(
    pegParse('< {level, select, A {1} 4 {2} 3 {3} 2{6} 1{12}} hours')
  ).toMatchSnapshot();
});

// https://github.com/formatjs/formatjs/issues/1845
test('greater than sign without escape', () => {
  expect(
    pegParse('> {level, select, A {1} 4 {2} 3 {3} 2{6} 1{12}} hours')
  ).toMatchSnapshot();
});

test('unmatched closing tag', () => {
  expect(() => pegParse('a </foo>')).toThrow();
});

test('unclosed opening tag', () => {
  expect(() => pegParse('<foo>a')).toThrow();
});

test('unclosed opening tag with nested messages', () => {
  expect(() =>
    pegParse(`
    You have <em>{count, plural, =1 {</em> one message} other {<b>#</b> messages}}.
  `)
  ).toThrow();
});
