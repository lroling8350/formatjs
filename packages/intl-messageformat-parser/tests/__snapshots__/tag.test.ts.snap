// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`escaped multiple tags 1`] = `
Array [
  Object {
    "type": 0,
    "value": "I <3 cats. <a>foo</a> <b>bar</b>",
  },
]
`;

exports[`escaped multiple tags with placeholder 1`] = `
Array [
  Object {
    "type": 0,
    "value": "I <3 cats. <a>foo</a> <b>",
  },
  Object {
    "type": 1,
    "value": "bar",
  },
  Object {
    "type": 0,
    "value": "</b>",
  },
]
`;

exports[`escaped tag with rich arg 1`] = `
Array [
  Object {
    "type": 0,
    "value": "I <3 cats.",
  },
]
`;

exports[`greater than sign without escape 1`] = `
Array [
  Object {
    "type": 0,
    "value": "> ",
  },
  Object {
    "options": Object {
      "1": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "12",
          },
        ],
      },
      "2": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "6",
          },
        ],
      },
      "3": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "3",
          },
        ],
      },
      "4": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "2",
          },
        ],
      },
      "A": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "1",
          },
        ],
      },
    },
    "type": 5,
    "value": "level",
  },
  Object {
    "type": 0,
    "value": " hours",
  },
]
`;

exports[`issue #1845: less than sign without escape 1`] = `
Array [
  Object {
    "type": 0,
    "value": "< ",
  },
  Object {
    "options": Object {
      "1": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "12",
          },
        ],
      },
      "2": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "6",
          },
        ],
      },
      "3": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "3",
          },
        ],
      },
      "4": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "2",
          },
        ],
      },
      "A": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "1",
          },
        ],
      },
    },
    "type": 5,
    "value": "level",
  },
  Object {
    "type": 0,
    "value": " hours",
  },
]
`;

exports[`mismatched tag with ignoreTag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is <a>mismatch</b>",
  },
]
`;

exports[`nested tag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is ",
  },
  Object {
    "children": Array [
      Object {
        "type": 0,
        "value": "nested ",
      },
      Object {
        "children": Array [
          Object {
            "type": 1,
            "value": "placeholder",
          },
        ],
        "type": 8,
        "value": "b",
      },
    ],
    "type": 8,
    "value": "a",
  },
]
`;

exports[`nested tag with ignoreTag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is <a>nested <b>",
  },
  Object {
    "type": 1,
    "value": "placeholder",
  },
  Object {
    "type": 0,
    "value": "</b></a>",
  },
]
`;

exports[`self-closing tag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is ",
  },
  Object {
    "type": 0,
    "value": "<br/>",
  },
  Object {
    "type": 0,
    "value": " ",
  },
  Object {
    "children": Array [
      Object {
        "type": 0,
        "value": "nested ",
      },
      Object {
        "children": Array [
          Object {
            "type": 1,
            "value": "placeholder",
          },
        ],
        "type": 8,
        "value": "b",
      },
    ],
    "type": 8,
    "value": "a",
  },
]
`;

exports[`self-closing tag with ignoreTag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is <br/> <a>nested <b>",
  },
  Object {
    "type": 1,
    "value": "placeholder",
  },
  Object {
    "type": 0,
    "value": "</b></a>",
  },
]
`;

exports[`tag in plural 1`] = `
Array [
  Object {
    "type": 0,
    "value": "You have ",
  },
  Object {
    "offset": 0,
    "options": Object {
      "=1": Object {
        "location": undefined,
        "value": Array [
          Object {
            "children": Array [
              Object {
                "type": 0,
                "value": "1",
              },
            ],
            "type": 8,
            "value": "b",
          },
          Object {
            "type": 0,
            "value": " Message",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "children": Array [
              Object {
                "type": 7,
              },
            ],
            "type": 8,
            "value": "b",
          },
          Object {
            "type": 0,
            "value": " Messages",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "count",
  },
]
`;

exports[`tag in plural with ignoreTag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "You have ",
  },
  Object {
    "offset": 0,
    "options": Object {
      "=1": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "<b>1</b> Message",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "<b>",
          },
          Object {
            "type": 7,
          },
          Object {
            "type": 0,
            "value": "</b> Messages",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "count",
  },
]
`;

exports[`tag with dash 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is ",
  },
  Object {
    "type": 0,
    "value": "<br/>",
  },
  Object {
    "type": 0,
    "value": " ",
  },
  Object {
    "children": Array [
      Object {
        "type": 0,
        "value": "nested ",
      },
      Object {
        "children": Array [
          Object {
            "type": 1,
            "value": "placeholder",
          },
        ],
        "type": 8,
        "value": "b",
      },
    ],
    "type": 8,
    "value": "dash-tag",
  },
]
`;

exports[`tag with dash with ignoreTag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is <br/> <dash-tag>nested <b>",
  },
  Object {
    "type": 1,
    "value": "placeholder",
  },
  Object {
    "type": 0,
    "value": "</b></dash-tag>",
  },
]
`;

exports[`tag with number arg 1`] = `
Array [
  Object {
    "type": 0,
    "value": "I have ",
  },
  Object {
    "children": Array [
      Object {
        "style": null,
        "type": 2,
        "value": "numCats",
      },
    ],
    "type": 8,
    "value": "foo",
  },
  Object {
    "type": 0,
    "value": " cats.",
  },
]
`;

exports[`tag with number arg with ignoreTag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "I have <foo>",
  },
  Object {
    "style": null,
    "type": 2,
    "value": "numCats",
  },
  Object {
    "type": 0,
    "value": "</foo> cats.",
  },
]
`;

exports[`tag with rich arg 1`] = `
Array [
  Object {
    "type": 0,
    "value": "I ",
  },
  Object {
    "children": Array [
      Object {
        "type": 0,
        "value": "have",
      },
    ],
    "type": 8,
    "value": "b",
  },
  Object {
    "type": 0,
    "value": " ",
  },
  Object {
    "children": Array [
      Object {
        "style": null,
        "type": 2,
        "value": "numCats",
      },
      Object {
        "type": 0,
        "value": " some string ",
      },
      Object {
        "type": 1,
        "value": "placeholder",
      },
    ],
    "type": 8,
    "value": "foo",
  },
  Object {
    "type": 0,
    "value": " cats.",
  },
]
`;

exports[`tag with rich arg with ignoreTag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "I <b>have</b> <foo>",
  },
  Object {
    "style": null,
    "type": 2,
    "value": "numCats",
  },
  Object {
    "type": 0,
    "value": " some string ",
  },
  Object {
    "type": 1,
    "value": "placeholder",
  },
  Object {
    "type": 0,
    "value": "</foo> cats.",
  },
]
`;

exports[`unescaped left angle bracket 1`] = `
Array [
  Object {
    "type": 0,
    "value": "I <3 cats.",
  },
]
`;
